ResearchKit/Common/ORKFormStepViewController.m

```diff
@@ -521,8 +521,19 @@ - (BOOL)allAnswered {
     return ([self numAnswered] == [self formItems].count);
 }
 
+- (BOOL)allAnswersValid {
+    for (ORKFormItem *item in [self formItems]) {
+        id answer = _savedAnswers[item.identifier];
+        BOOL isNonNull = answer && ![answer isKindOfClass:[NSNull class]];
+        if (isNonNull && ![item.impliedAnswerFormat isAnswerValid:answer]) {
+            return NO;
+        }
+    }
+    return YES;
+}
+
 - (BOOL)continueButtonEnabled {
-    return ([self allAnswered] || (self.step.optional && ([self numAnswered] > 0 || ! self.skipButtonItem)));
+    return ([self allAnswersValid] && ([self allAnswered] || (self.step.optional && ([self numAnswered] > 0 || ! self.skipButtonItem))));
```

Comment:
A little hard to read, how about splitting this one line to something like:

```
BOOL optionalButNotEmpty = self.step.optional && ([self numAnswered] > 0 || ! self.skipButtonItem);
return [self allAnswersValid] && ([self allAnswered] || optionalButNotEmpty);
```

And put some documentation around that?